### PR TITLE
Multipart Form Params - Extract boundary from body if not in header

### PR DESCRIPTION
**Pull request by kingthorin**
- VariantMultipartFormParameters > Add handling to get attempt to get the boundary from the body if not defined in the content-type header.
- VariantMultipartFormParametersUnitTest > Add tests to assert the new functionality.

Related to: zaproxy/zap-extensions#5024

### CODE REVIEW

File: zap/src/main/java/org/parosproxy/paros/core/scanner/VariantMultipartFormParameters.java [hunk 1]
```diff
@@ -80,13 +80,17 @@
...
         int position = 0;
         int offset = 0;
-        String boundary = getBoundary(contentType) + HttpHeader.CRLF;
-        for (String part : msg.getRequestBody().toString().split(boundary)) {
+        String bareBoundary = getBoundary(contentType, msg);
+        if (bareBoundary == null) {
+            return;
+        }
+        String boundaryCrlf = bareBoundary + HttpHeader.CRLF;
+        for (String part : msg.getRequestBody().toString().split(boundaryCrlf)) {
```
[the remark this answers is not in the document]
Comment (thc202) [reply zap/src/main/java/org/parosproxy/paros/core/scanner/VariantMultipartFormParameters.java:88]: Needs pattern quote.

**Comment by kingthorin** [comment]
🙂 ping 🙂

File: zap/src/main/java/org/parosproxy/paros/core/scanner/VariantMultipartFormParameters.java [hunk 2]
```diff
@@ -271,10 +275,10 @@ private String setParameter(
         return newBody;
     }
 
-    private String getBoundary(String contentTypeHeader) {
+    private String getBoundary(String contentTypeHeader, HttpMessage msg) {
```
Comment (thc202) [inline zap/src/main/java/org/parosproxy/paros/core/scanner/VariantMultipartFormParameters.java:278]: `static`

File: zap/src/main/java/org/parosproxy/paros/core/scanner/VariantMultipartFormParameters.java [hunk 3]
```diff
@@ -286,4 +290,12 @@ private String getBoundary(String contentTypeHeader) {
 
         return boundary;
     }
+
+    private String getBoundaryFromBody(HttpMessage msg) {
```
Comment (thc202) [inline zap/src/main/java/org/parosproxy/paros/core/scanner/VariantMultipartFormParameters.java:294]: `static`

File: zap/src/main/java/org/parosproxy/paros/core/scanner/VariantMultipartFormParameters.java [hunk 4]
```diff
@@ -286,4 +290,12 @@ private String getBoundary(String contentTypeHeader) {
...
+        String body = msg.getRequestBody().toString();
+        if (body.startsWith("--")) {
```
Comment (thc202) [inline zap/src/main/java/org/parosproxy/paros/core/scanner/VariantMultipartFormParameters.java:296]: Needs a test where the body is exactly `--`.

File: zap/src/test/java/org/parosproxy/paros/core/scanner/VariantMultipartFormParametersUnitTest.java [hunk 5]
```diff
@@ -112,6 +112,58 @@ void shouldExtractParametersFromAllParts() {
...
+        assertThat(variant.getParamList().get(3).getName(), is(equalTo("somefile")));
+        assertThat(
+                variant.getParamList().get(3).getValue(), is(equalTo(DEFAULT_FILE_PARAM_CONTENT)));
+        assertThat(
+                variant.getParamList().get(3).getType(),
+                is(equalTo(NameValuePair.TYPE_MULTIPART_DATA_FILE_PARAM)));
+    }
+
+    @Test
+    void shouldNotExtractParametersIfBoundaryNotidentified() {
```
Comment (thc202) [inline zap/src/test/java/org/parosproxy/paros/core/scanner/VariantMultipartFormParametersUnitTest.java:155]: `Identified`

File: zap/src/main/java/org/parosproxy/paros/core/scanner/VariantMultipartFormParameters.java (hunk 4 above)
Comment (thc202) [reply zap/src/main/java/org/parosproxy/paros/core/scanner/VariantMultipartFormParameters.java:296]: Also worth extracting a constant.

**Comment by kingthorin** [comment]
Thanks

**Comment by kingthorin** [comment]
Now rebased and addressed review.

File: zap/src/main/java/org/parosproxy/paros/core/scanner/VariantMultipartFormParameters.java [hunk 6]
```diff
@@ -73,20 +75,25 @@ public void setMessage(HttpMessage msg) {
         }
 
         String contentType = msg.getRequestHeader().getHeader(HttpHeader.CONTENT_TYPE);
-        if (contentType == null || !contentType.toLowerCase().startsWith("multipart/form-data")) {
+        if (contentType == null
+                || !contentType.toLowerCase(Locale.ROOT).startsWith("multipart/form-data")) {
```
Comment (kingthorin) [inline zap/src/main/java/org/parosproxy/paros/core/scanner/VariantMultipartFormParameters.java:79]: Added `Locale.ROOT` just as a "clean code" action.

Comment (thc202) [reply zap/src/main/java/org/parosproxy/paros/core/scanner/VariantMultipartFormParameters.java:79]: It's actually a bug, better use `HttpHeader.hasContentType(String)` or `getNormalisedContentTypeValue()`

Comment (thc202) [reply zap/src/main/java/org/parosproxy/paros/core/scanner/VariantMultipartFormParameters.java:79]: Worth adding a test.

Comment (kingthorin) [reply zap/src/main/java/org/parosproxy/paros/core/scanner/VariantMultipartFormParameters.java:79]: Happy to, but I don't really know ow the condition(s) when the missing Locale comes into play.

Comment (thc202) [reply zap/src/main/java/org/parosproxy/paros/core/scanner/VariantMultipartFormParameters.java:79]: Make the content-type all upper case and set the default locale to Turkish (e.g. `Locale.setDefault(new Locale.Builder().setLanguageTag("tr").build());`.

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thank you!

**Review by psiinon — APPROVED** [review]

**Pull request merged**